### PR TITLE
hal_nordic: Rework nRF IEEE 802.15.4 temperature update module

### PR DESCRIPTION
**Pull request by pkwiek**
This commit reworks nRF IEEE 802.15.4 temperature update module to save
resources, otherwise used to create thread, by using work queue.

### CODE REVIEW

File: modules/hal_nordic/nrf_802154/sl_opensource/platform/nrf_802154_temperature_zephyr.c [hunk 1]
```diff
@@ -18,66 +18,56 @@
...
-	temperature_dev = DEVICE_DT_GET(DT_NODELABEL(temp));
+	k_work_reschedule(&temperature.work, K_MSEC(CONFIG_NRF_802154_TEMPERATURE_UPDATE_PERIOD))
+}
 
-	if (!device_is_ready(temperature_dev)) {
+void nrf_802154_temperature_init(void)
+{
+	temperature.dev = DEVICE_DT_GET(DT_NODELABEL(temp));
+
+	if (!device_is_ready(temperature.dev)) {
```
Comment (pdunaj) [inline modules/hal_nordic/nrf_802154/sl_opensource/platform/nrf_802154_temperature_zephyr.c:56]: asserts are turned off by default
if the code won't work when the device is not ready the error handler should emit an error and return. Assert can be left there to simplify debugging.

If we expect device is always ready (i.e. this situation cannot ever occur) there should be no error code, simply assert in form
```
__ASSERT_NO_MSG(device_is_ready(temperature.dev));
```

File: modules/hal_nordic/nrf_802154/sl_opensource/platform/nrf_802154_temperature_zephyr.c [hunk 2]
```diff
@@ -18,66 +18,56 @@
...
 {
-	/* Intentionally empty. */
-}
+	struct sensor_value temperature_value;
 
-int8_t nrf_802154_temperature_get(void)
-{
-	return temperature;
-}
+	sensor_sample_fetch(temperature.dev);
```
Comment (pdunaj) [inline modules/hal_nordic/nrf_802154/sl_opensource/platform/nrf_802154_temperature_zephyr.c:40]: please check values returned from fetch and get.

**Review by pdunaj — CHANGES_REQUESTED** [review]

File: modules/hal_nordic/nrf_802154/sl_opensource/platform/nrf_802154_temperature_zephyr.c (hunk 2 above)
Comment (pkwiek) [reply modules/hal_nordic/nrf_802154/sl_opensource/platform/nrf_802154_temperature_zephyr.c]: Values are checked and only reschedule is performed if fet of get failed.

File: modules/hal_nordic/nrf_802154/sl_opensource/platform/nrf_802154_temperature_zephyr.c (hunk 1 above)
Comment (pkwiek) [reply modules/hal_nordic/nrf_802154/sl_opensource/platform/nrf_802154_temperature_zephyr.c]: We do not expect die temperature sensor would malfunction, hence only an assert is left.

File: modules/hal_nordic/nrf_802154/sl_opensource/platform/nrf_802154_temperature_zephyr.c [hunk 3]
```diff
@@ -18,66 +18,68 @@
...
-	/* Intentionally empty. */
-}
+	struct k_work_delayable work;
+} temperature;
 
-void nrf_802154_temperature_deinit(void)
+static void temperature_update_reschedule(void)
 {
-	/* Intentionally empty. */
+	k_work_reschedule(&temperature.work, K_MSEC(CONFIG_NRF_802154_TEMPERATURE_UPDATE_PERIOD))
```
Comment (emob-nordic) [inline modules/hal_nordic/nrf_802154/sl_opensource/platform/nrf_802154_temperature_zephyr.c:38]: `;` missed at the end of line. It doesn't compile.

File: modules/hal_nordic/nrf_802154/sl_opensource/platform/nrf_802154_temperature_zephyr.c [hunk 4]
```diff
@@ -18,66 +18,68 @@
...
+	temperature.dev = DEVICE_DT_GET(DT_NODELABEL(temp));
 
-		if (temperature != temperature_value.val1) {
-			temperature = temperature_value.val1;
+	__ASSERT_NO_MSG(device_is_ready(temperature.dev));
 
-			nrf_802154_temperature_changed();
-		}
+	k_work_init_delayable(&temperature.work, work_handler);
+	k_delayed_work_submit(&temperature.work, K_NO_WAIT);
```
Comment (emob-nordic) [inline modules/hal_nordic/nrf_802154/sl_opensource/platform/nrf_802154_temperature_zephyr.c:74]: `k_delayed_work_submit` is depracated, use `k_work_schedule` instead.

File: modules/hal_nordic/nrf_802154/sl_opensource/platform/nrf_802154_temperature_zephyr.c [hunk 5]
```diff
@@ -18,66 +18,68 @@
...
+	temperature_update_reschedule();
+}
+
+void nrf_802154_temperature_init(void)
+{
+	temperature.dev = DEVICE_DT_GET(DT_NODELABEL(temp));
 
-		if (temperature != temperature_value.val1) {
-			temperature = temperature_value.val1;
+	__ASSERT_NO_MSG(device_is_ready(temperature.dev));
```
Comment (emob-nordic) [inline modules/hal_nordic/nrf_802154/sl_opensource/platform/nrf_802154_temperature_zephyr.c:71]: Our sample hits this assert.

**Review by eobalski — CHANGES_REQUESTED** [review]

**Review by eobalski — APPROVED** [review]
Maybe not the most beautiful, but works fine. Thanks!

File: modules/hal_nordic/nrf_802154/sl_opensource/platform/nrf_802154_temperature_zephyr.c [hunk 6]
```diff
@@ -18,66 +18,92 @@
...
-		sensor_channel_get(temperature_dev, SENSOR_CHAN_DIE_TEMP, &temperature_value);
+	return 0;
+}
+#endif /* defined(CONFIG_NRF_802154_TEMPERATURE_UPDATE) */
 
-		if (temperature != temperature_value.val1) {
-			temperature = temperature_value.val1;
+void nrf_802154_temperature_init(void)
+{
+#if defined(CONFIG_NRF_802154_TEMPERATURE_UPDATE)
```
Comment (pdunaj) [inline modules/hal_nordic/nrf_802154/sl_opensource/platform/nrf_802154_temperature_zephyr.c:88]: this should be handled in header - there is no point making a call if this does not exist

File: modules/hal_nordic/nrf_802154/sl_opensource/platform/nrf_802154_temperature_zephyr.c [hunk 7]
```diff
@@ -18,66 +18,92 @@
...
-K_THREAD_DEFINE(temperature_update_tid,
-		CONFIG_NRF_802154_TEMPERATURE_UPDATE_STACK_SIZE,
-		temperature_update_thread,
-		NULL,
-		NULL,
-		NULL,
-		CONFIG_NRF_802154_TEMPERATURE_UPDATE_PRIO,
-		0,
-		0);
+#if defined(CONFIG_NRF_802154_TEMPERATURE_UPDATE)
```
Comment (pdunaj) [inline modules/hal_nordic/nrf_802154/sl_opensource/platform/nrf_802154_temperature_zephyr.c:105]: this should be handled in cmake

**Review by pdunaj — CHANGES_REQUESTED** [review]

**Review by pdunaj — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**